plutus-ledger/src/Ledger/Index/Internal.hs

```diff
@@ -54,4 +87,63 @@ deriving via (PrettyShow ValidationError) instance Pretty ValidationError
 data ValidationPhase = Phase1 | Phase2 deriving (Eq, Show, Generic, FromJSON, ToJSON)
 deriving via (PrettyShow ValidationPhase) instance Pretty ValidationPhase
 type ValidationErrorInPhase = (ValidationPhase, ValidationError)
-type ValidationSuccess = Map.Map RdmrPtr ([Text], ExUnits)
+type ValidationSuccess = (RedeemerReport, Validated (Tx EmulatorEra))
+type RedeemerReport = Map.Map RdmrPtr ([Text], ExUnits)
+
+data ValidationResult
+    = FailPhase1 !CardanoTx !ValidationError
+    -- ^ A transaction failed to validate in phase 1.
+    | FailPhase2 !OnChainTx !ValidationError !C.Value
+    -- ^ A transaction failed to validate in phase 2. The @Value@ indicates the amount of collateral stored in the transaction.
+    | Success !OnChainTx !RedeemerReport
+    deriving stock (Eq, Show, Generic)
+
+makePrisms ''ValidationResult
+
+data ValidationResultSimple
+    = ValidationFailPhase1 !CardanoTx !ValidationError
+    | ValidationFailPhase2 !CardanoTx !ValidationError !C.Value
+    | ValidationSuccess !CardanoTx
+    deriving stock Generic
+    deriving anyclass (ToJSON, FromJSON)
+
+toValidationResultSimple :: ValidationResult -> ValidationResultSimple
+toValidationResultSimple (FailPhase1 tx err)      = ValidationFailPhase1 tx err
+toValidationResultSimple (FailPhase2 vtx err val) = ValidationFailPhase2 (unOnChain vtx) err val
+toValidationResultSimple (Success vtx _)          = ValidationSuccess (unOnChain vtx)
+
+instance ToJSON ValidationResult where
+    toJSON = toJSON . toValidationResultSimple
+
+instance FromJSON ValidationResult where
+    parseJSON = mempty -- Always fail, this instance isn't really used, but required by pab's logging framework.
+
+instance Pretty ValidationResult where
+    pretty res = hang 2 $ vsep $
+        case res of
+            FailPhase1 _ err   -> "Validation failed in phase 1:" <+> pretty err
+            FailPhase2 _ err _ -> "Validation failed in phase 2:" <+> pretty err
```

Comment:
Do you think it be worth having a golden test for phase 1 and phase 2 failures?

Reply:
Absolutely! Currently we test this somewhat indirectly with the `checkErrorWhitelist` tests, but I don't understand them well, so I'm actually not sure if our current behaviour regarding phase1 and phase2 error is correct!